modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCutManager.java

```diff
@@ -0,0 +1,217 @@
+        super.start0();
+
+        cctx.exchange().registerExchangeAwareComponent(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void stop0(boolean cancel) {
+        cancelCut(new IgniteCheckedException("Ignite node is stopping."));
+    }
+
+    /**
+     * Stops Consistent Cut in case of baseline topology changed.
+     */
+    @Override public void onInitBeforeTopologyLock(GridDhtPartitionsExchangeFuture fut) {
+        if (fut.changedBaseline() || fut.isBaselineNodeFailed())
+            cancelCut(new IgniteCheckedException("Ignite topology changed, can't finish Consistent Cut."));
+    }
+
+    /**
+     * Registers transaction before it is removed from {@link IgniteTxManager#activeTransactions()}.
+     *
+     * @param tx Transaction.
+     * @param firstCommit If {@code true} then it is first commit for distributed transaction.
+     */
+    public void onRemoveActiveTransaction(IgniteInternalTx tx, boolean firstCommit) {
+        ConsistentCutFuture cut = cutFutRef.get();
+
+        if (cut != null) {
+            if (firstCommit)
+                tx.cutId(cut.id());
+
+            cut.onRemoveActiveTransaction(tx.finishFuture());
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Register transaction [nearXid=" + tx.nearXidVersion().asIgniteUuid()
+                + ", xid=" + tx.xid() + ", txCutId=" + tx.cutId() + ", cutId=" + (cut == null ? null : cut.id()) + ']');
+        }
+    }
+
+    /**
+     * Wraps a transaction message if Consistent Cut is running.
+     *
+     * @param txMsg Transaction message to wrap.
+     * @param txCutId Consistent Cut ID after which transaction committed, if specified.
+     */
+    public static GridCacheMessage wrapMessage(
+        GridCacheSharedContext<?, ?> cctx,
+        GridCacheMessage txMsg,
+        @Nullable UUID txCutId
+    ) {
+        if (cctx.consistentCutMgr() == null)
+            return txMsg;
+
+        ConsistentCutFuture cut = cctx.consistentCutMgr().cutFuture();
+
+        if (cut != null)
+            return new ConsistentCutAwareMessage(txMsg, cut.id(), txCutId);
+
+        return txMsg;
+    }
+
+    /**
+     * Handles received Consistent Cut ID from remote node. It compares it with the latest ID that local node is aware of.
+     * Init local Consistent Cut procedure if received ID is a new one.
+     *
+     * @param id ID of {@link ConsistentCutFuture}.
+     */
+    public void handleConsistentCutId(UUID id) {
+        ConsistentCutFuture cut = cutFutRef.get();
+
+        if (cut != null || Objects.equals(id, lastFinishedCutId))
+            return;
+
+        ConsistentCutFuture newCut = newConsistentCut(id);
+
+        if (!cutFutRef.compareAndSet(cut, newCut))
+            return;
+
+        if (cctx.kernalContext().clientNode() ||
+            !baselineNode(cctx.localNode(), cctx.kernalContext().state().clusterState())
+        ) {
+            newCut.onDone();
```

Comment:
We need to create `GridFinishedFuture` in this case.
No need to create `ConsistentCutFuture` object here.

Reply:
Fix, but in this case GridFinishedFuture must hold `id`. Make a separate class for this reason.